### PR TITLE
sg: add a fallback func for secrets if we fail fetching a secret

### PR DESCRIPTION
**Pull request by burmudar**
Add a fallback function when we fail fetching a secret from our primary provider. 

## Test plan
Tested locally

### CODE REVIEW

File: dev/sg/internal/bk/bk.go [hunk 1]
```diff
@@ -94,9 +95,16 @@ func NewClient(ctx context.Context, out *std.Output) (*Client, error) {
 	if err != nil {
 		return nil, err
 	}
+	// Get the secret from gcloud otherwise as fallback try getting it from the environment
```
Comment (jhchabran) [inline dev/sg/internal/bk/bk.go:98]: In the context of this specific secret, shouldn't we get it first from the environment?

File: dev/sg/internal/secrets/store.go [hunk 2]
```diff
@@ -118,7 +120,7 @@ func (s *Store) Get(key string, target any) error {
 	return errors.Newf("%w: %s not found", ErrSecretNotFound, key)
 }
 
-func (s *Store) GetExternal(ctx context.Context, secret ExternalSecret) (string, error) {
+func (s *Store) GetExternal(ctx context.Context, secret ExternalSecret, fallback FallbackFunc) (string, error) {
```
Comment (jhchabran) [inline dev/sg/internal/secrets/store.go:123]: What about having the options pattern here? passing nil all the time except once may be a bit annoying.

Comment (burmudar) [reply dev/sg/internal/secrets/store.go:123]: Yeah. I honestly think that is much better. I am a bit annoyed with the current API atm so thought I get some feedback first :)

File: dev/sg/internal/bk/bk.go (hunk 1 above)
Comment (burmudar) [reply dev/sg/internal/bk/bk.go]: there was an unused `retrieveToken` func which did this, which I am now using again :)

File: dev/sg/internal/secrets/store.go [hunk 3]
```diff
@@ -118,7 +120,18 @@ func (s *Store) Get(key string, target any) error {
 	return errors.Newf("%w: %s not found", ErrSecretNotFound, key)
 }
 
-func (s *Store) GetExternal(ctx context.Context, secret ExternalSecret) (string, error) {
+func EnvFallback(ctx context.Context, envKey string) FallbackFunc {
```
Comment (jhchabran) [inline dev/sg/internal/secrets/store.go:123]: I'm hesitant on this one, it feels like it should always be env ovverides rather than fallback :/

Comment (burmudar) [reply dev/sg/internal/secrets/store.go:123]: I'm fine with removing it. It isn't being used actually

**Review by jhchabran — APPROVED** [review]
Let's drop the risky fallback and we're good :)

File: dev/sg/internal/secrets/store.go (hunk 3 above)
Comment (sanderginn) [reply dev/sg/internal/secrets/store.go]: This doesn't give Buildkite a fallback for getting the env var, does it?

Comment (burmudar) [reply dev/sg/internal/secrets/store.go]: Nope - I think how cases like that should be handled is like how it is handled in `getToken` where the env is checked before asking the secret store.

I do have an idea of an api that could support checking the env first and then ask something else, but that will require more time to do it

**Pull request merged**